File: leetcode/problems/cpp/isomorphic-strings.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(1)

#include "string"
using namespace std;

class Solution {
  public:
   bool isIsomorphic(string s, string t) {
      if (s.size() != t.size()) return false;

      char s_map[256] = {0};
      char t_map[256] = {0};

      for (int i = 0; i < s.size(); i++) {
         if (s_map[s[i]] != t_map[t[i]]) return false;
         s_map[s[i]] = i + 1;
         t_map[t[i]] = i + 1;
      }

      return true;
   }
};

File: leetcode/problems/rs/isomorphic-strings.rs
// by @codeAbinash
// Time : O(n)
// Space : O(1)

impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        let mut s_map = [0; 128];
        let mut t_map = [0; 128];
        let mut index = 1;

        for (s_char, t_char) in s.chars().zip(t.chars()) {
            let s_char = s_char as usize;
            let t_char = t_char as usize;

            if s_map[s_char] != t_map[t_char] {
                return false;
            }

            if s_map[s_char] == 0 {
                s_map[s_char] = index;
                t_map[t_char] = index;
                index += 1;
            }
        }
        true
    }
}
